common/transport/amqp/src/amqp.ts

```diff
@@ -454,232 +485,60 @@ export class Amqp {
-                // operations, accept it.  This could be a put token that we previously
-                // timed out.  Be happy.  It made it home, just too late to be useful.
-                //
-                /*Codes_SRS_NODE_COMMON_AMQP_06_012: [All responses shall be completed.]*/
-                this._receivers[_putTokenReceivingEndpoint].complete(msg);
-              });
-              /*Codes_SRS_NODE_COMMON_AMQP_06_020: [If given as an argument, the `initializeCBS` method shall call `initializeCBSCallback` with a null error object if successful.]*/
-              this._safeCallback(initializeCBSCallback, null);
-            }
-          });
-        }
-      });
-    }
-  }
-
-  private _removeExpiredPutTokens(): void {
-    const currentTime = Math.round(Date.now() / 1000);
-    let expiredPutTokens: PutTokenOperation[] = [];
-    while (this._putToken.outstandingPutTokens.length > 0) {
-      //
-      // The timeouts in this array by definition are monotonically increasing.  We will be done looking if we
-      // hit one that is not yet expired.
-      //
-      /*Codes_SRS_NODE_COMMON_AMQP_06_007: [ The `putToken` method will time out the put token operation if no response is returned within a configurable number of seconds.]*/
-      if (this._putToken.outstandingPutTokens[0].expirationTime < currentTime) {
-        expiredPutTokens.push(this._putToken.outstandingPutTokens[0]);
-        this._putToken.outstandingPutTokens.splice(0, 1);
-      } else {
-        break;
+    for (let receiverEndpoint in this._receivers) {
+      if (this._receivers.hasOwnProperty(receiverEndpoint)) {
+        remainingLinks.push(this._receivers[receiverEndpoint]);
+          delete this._receivers[receiverEndpoint];
       }
     }
-    expiredPutTokens.forEach((currentExpiredPut) => {
-      /*Codes_SRS_NODE_COMMON_AMQP_06_008: [ The `putToken` method will invoke the `putTokenCallback` (if supplied) with an error object if the put token operation timed out. .]*/
-      this._safeCallback(currentExpiredPut.putTokenCallback, new errors.TimeoutError('Put Token operation had no response within ' + this._putToken.numberOfSecondsToTimeout));
-    });
-    //
-    // If there are any putTokens left keep trying to time them out.
-    //
-    if (this._putToken.outstandingPutTokens.length > 0) {
-      this._putToken.timeoutTimer = setTimeout(this._removeExpiredPutTokens.bind(this), this._putToken.putTokenTimeOutExaminationInterval);
-    }
+
+    return remainingLinks;
   }
 
   /*Codes_SRS_NODE_COMMON_AMQP_16_011: [All methods should treat the `done` callback argument as optional and not throw if it is not passed as argument.]*/
-  private _safeCallback(callback: GenericAmqpBaseCallback, error?: Error | null, result?: any): void {
+  private _safeCallback(callback: GenericAmqpBaseCallback<any>, error?: Error | null, result?: any): void {
     if (callback) {
       process.nextTick(() => callback(error, result));
```

Comment:
>process.nextTick [](start = 6, length = 16)

uh oh, this suggests a work-around and scares me.  Why do you do this?  All callbacks lose their stack and that bothers me.
If this must be here, suggest a better function name like _deferredSafeCallback() so the caller knows what's going on (I read this whole file without knowing that the nextTick call was in here, and I'm worried that this was significant in understanding how the code functions.)

Reply:
the problem when calling callbacks directly from promises is that any error in one of those callbacks end up getting caught by the .catch() clause of said promise, which screws up the stack way more than using process.nextTick. happy to hear suggestions though

---
In reply to: [139535307](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139535307) [](ancestors = 139535307)

Reply:
Interesting.  I learned something new today,.  I'll have to think about the implications on .catch more.

---
In reply to: [139561431](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139561431) [](ancestors = 139561431,139535307)